components/feature/download/src/main/java/mozilla/components/feature/download/DownloadManager.kt

```diff
@@ -0,0 +1,139 @@
+import android.content.Context
+import android.net.Uri
+import android.support.annotation.RequiresPermission
+import mozilla.components.browser.session.Download
+import mozilla.components.support.ktx.android.content.isPermissionGranted
+import mozilla.components.support.utils.DownloadUtils
+import android.content.BroadcastReceiver
+import android.content.Intent
+import android.content.IntentFilter
+import java.lang.IllegalArgumentException
+import android.app.DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED
+import android.app.DownloadManager.Request
+import android.app.DownloadManager.EXTRA_DOWNLOAD_ID
+import android.app.DownloadManager.ACTION_DOWNLOAD_COMPLETE
+import android.content.Context.DOWNLOAD_SERVICE
+
+typealias OnDownloadFinish = (Long) -> Unit
+typealias AndroidDownloadManager = android.app.DownloadManager
+
+/**
+ * Handles the interactions with the [AndroidDownloadManager].
+ * @property context A reference to [Context] applicationContext.
+ */
+class DownloadManager(private val context: Context) {
+
+    private val queuedDownloads = HashMap<Long, OnDownloadFinish>()
+    private var isSubscribedReceiver = false
+    private lateinit var androidDownloadManager: AndroidDownloadManager
+
+    /**
+     * Schedule a download through the [AndroidDownloadManager].
+     * @param download metadata related to the download.
+     * @param downloadListener a listener to be notify when the download is completed.
+     * @param refererURL the url from where this download was referred.
+     * @param cookie any additional cookie to add as part of the download request.
+     * @return the id reference of the scheduled download.
+     */
+    @RequiresPermission(allOf = arrayOf(INTERNET, WRITE_EXTERNAL_STORAGE))
+    fun download(
+        download: Download,
+        downloadListener: OnDownloadFinish = {},
+        refererURL: String = "",
+        cookie: String = ""
+    ): Long {
+
+        if (!context.isPermissionGranted(INTERNET, WRITE_EXTERNAL_STORAGE))
+            throw SecurityException("You must be granted INTERNET and WRITE_EXTERNAL_STORAGE permissions")
+
+        androidDownloadManager = context.getSystemService(DOWNLOAD_SERVICE) as AndroidDownloadManager
+
+        val fileName = getFileName(download)
+
+        val request = Request(Uri.parse(download.url))
+            .setNotificationVisibility(VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
+            .setMimeType(download.contentType)
+            .addRequestHeader("User-Agent", download.userAgent)
+
+        if (refererURL.isNotEmpty())
+            request.addRequestHeader("Cookie", cookie)
+
+        val downloadID: Long
+
+        request.setDestinationInExternalPublicDir(download.destinationDirectory, fileName)
+
+        downloadID = androidDownloadManager.enqueue(request)
+        queuedDownloads[downloadID] = downloadListener
+
+        if (!isSubscribedReceiver)
+            registerBroadcastReceiver(context)
```

Comment:
nit: Let's always wrap blocks in `{ }`. We may want to enable a detekt rule for that.